hearth/lib/hearth/middleware/endpoint.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+# in Hearth middleware
+module Hearth
+  module Middleware
+    # Resolves endpoints from endpoint parameters and modifies
+    # the request with resolved endpoint, headers and auth schemes.
+    # @api private
+    class Endpoint
+      # @param [Class] app The next middleware in the stack.
+      # @param [#resolve_endpoint(endpoint_params)] endpoint_provider An object
+      #   that responds to a `resolve_endpoint(endpoint_params)` method
+      #   where `endpoint_params` is a service specific struct.
+      #   The method must return an {Hearth::RulesEngine::Endpoint} object.
+      # @param [#build(config, input, context)] param_builder An object that
+      #   responds to a `build(config, input, context)` method and returns
+      #   an endpoint_params object.
+      def initialize(
+        app, endpoint_provider:, param_builder:, **kwargs
+      )
+        @app = app
+        @param_builder = param_builder
+        @endpoint_provider = endpoint_provider
+        @config = kwargs
+      end
+
+      def call(input, context)
+        params = @param_builder.build(@config, input, context)
+        endpoint = @endpoint_provider.resolve_endpoint(params)
+        update_request(context, endpoint)
+
+        # TODO: combine with auth middleware
```

Comment:
We set an auth on context I believe. Maybe schemes also belongs under there somehow.

Reply:
Yeah, somehow the properties of matchign auth schemes need to "override" or be merged onto the already resovled auth on context (eg: for sigv4, replacing the region with signignRegion).